processor/resourcedetectionprocessor/internal/consul/metadata.go

```diff
@@ -0,0 +1,90 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consul
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/consul/api"
+)
+
+type consulMetadata interface {
+	// Hostname returns the OS hostname
+	Hostname(context.Context) (string, error)
+
+	// OSType returns the host operating system
+	Datacenter(context.Context) (string, error)
+
+	NodeID(context.Context) (string, error)
+
+	Meta(context.Context) (map[string]string, error)
+}
+
+type consulMetadataImpl struct {
+	consulClient *api.Client
+}
+
+func newConsulMetadata(config *api.Config) (consulMetadata, error) {
+	cli, err := api.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize consul client: %w", err)
+	}
+	return &consulMetadataImpl{consulClient: cli}, nil
+}
+
+func (d *consulMetadataImpl) Hostname(ctx context.Context) (string, error) {
+	info, err := d.consulClient.Agent().Self()
```

Comment:
While it's not a huge deal since resource detection runs once at startup, I think we should still ensure that `Self()` is only called once and read all the fields from that value, rather than calling it multiple times. I don't know if we really need this `metadata.go`, we could just initialize the API client and call `Self` in the detector to make that simpler

Reply:
Wanted to keep it idomatic with the rest of the providers so I refactored to only using a single `Metadata` function like the azure provider. Let me know you thoughts and I can refactor again